Kind error fix issue #110

rupertford opened the pull request:
Includes new exception and its use in a class. If this looks reasonable then I suggest we follow this approach for the other classes (as we modify them).

Also added an x-failing test in the wrong file (r404) as the error happens there (but due to errors earlier).

rupertford commented on doc/developers_guide.rst:
```diff
@@ -452,10 +452,16 @@ In this way fparser2 captures the `R202` rule hierarchy in its
+is to allow external tools to more simply manage fparser if it is used
+as a library.
+
+Eacho of the exceptions are now discussed in turn.
```
Oopso.

rupertford commented on src/fparser/two/Fortran2003.py:
```diff
@@ -689,32 +690,91 @@ def match(string):
+            raise InternalError(
+                "String argument '{0}' in class Kind_Selector method "
+                "match() is too short to be valid.".format(string))
+        if string[0].isspace() or string[-1].isspace():
```
OK, I've done that. As discussed we will probably need to support trailing and leading spaces in many other match() methods so I'll add it as something to check when reviewing on the wiki once this PR is accepted.

rupertford commented on src/fparser/two/Fortran2003.py:
```diff
@@ -689,32 +690,91 @@ def match(string):
+        :return: `None` if there is no match, otherwise a `tuple` of
+                 size 3 containing a '(', a single `list` which
+                 contains an instance of classes that have matched and
+                 a ')'.
```
Quite right.

rupertford commented on src/fparser/two/tests/fortran2003/test_kindselector_r404.py:
```diff
@@ -0,0 +1,262 @@
+    '''Test that None is returned if the left bracket is missing.'''
+    reader = get_reader("KIND = some_kind)")
+    ast = Kind_Selector(reader)
+    assert not ast
```
Well I like it as it is :-)

rupertford commented:
Addressed all comments. Back to @arporter for 2nd review.

arporter commented on src/fparser/scripts/fparser2.py:
```diff
@@ -102,6 +102,9 @@ def runner(_, options, args):
             return
+        except InternalError as msg:
+            print ("Internal error in fparser: {0}".format(str(msg)))
+            return
```
The modifications to this file are not covered by the test suite. Also, rather than just `return` as if everything is fine, should we not `exit(1)` instead?

rupertford commented:
I've hopefully addressed the reviewers comments. I've taken the liberty of fully covering the fparser2.py script file and have added some basic documentation. Ready for 3rd review from @arporter.

arporter commented on src/fparser/two/tests/test_scripts.py:
```diff
@@ -0,0 +1,210 @@
+    # runner() as argument options
+    from fparser.scripts import fparser2
+
+    class DummyArgs(object):
```
You use this class a lot so I suggest promoting it up into this test module and then re-using it as required.

arporter commented on src/fparser/two/tests/test_scripts.py:
```diff
@@ -0,0 +1,210 @@
+
+
+def test_runner_non_existant_file(capsys):
+    '''Test that the script deals with no files provided as expected.'''
```
This docstring does not describe this test :-)

arporter commented on src/fparser/two/tests/test_scripts.py:
```diff
@@ -84,13 +83,13 @@ def test_runner_set_mode(tmpdir, capsys):
+    # runner() as an argument options class
 
-    class DummyArgs(object):
+    class DummyArgsFree(object):
```
I would have done:

    dargs = DummyArgs()
    dargs.mode = "free"
    ...
but I think we've been through enough iterations already :-)

rupertford replied:
You can't change it if you like :-)
